### DISCUSSION TITLE
Marketing Groovy, GORM & Grails

### DISCUSSION THREAD

**Discussion by gsartori** [apache/grails-core#13559] [General]
Searching the web it looks like we are dead (or dying). Just look at the dates on any Grails related searches.
Groovy is a bit better thanks to Paul King's blog posts.

My opinion is that part of the Grails project is to communicate Groovy, GORM and Grails itself.

I'm opening this discussion as a way to track marketing ideas, proposals and actions to improve the communication of Groovy, GORM and Grails.

**Comment by gsartori** [comment]
IntelliJ has an open-source page reporting all the most important web frameworks/languages foundations. 
Groovy & Grails is not one of those.

I have no idea about what it takes to be there, but that's a promotional page that can give confidence to people to adopt Grails (or at least give it a try) (or at least be aware of).

Should we be there with the other cool guys?

![Screenshot 2024-07-18 at 13 53 03](https://github.com/user-attachments/assets/478e6ada-5563-420f-aafd-145f04b8739c)

**Reply by snimavat** [reply]
We/Grails was once the "cool guy" on the street

**Comment by kico-itexto** [comment]
I recently asked on Reddit about the relevance of Groovy today. The .aí. feeling is that yes, we are dying.



https://www.reddit.com/r/groovy/s/xKypXEX6va

**Reply by kico-itexto** [reply]
Is GORM relevant outside the Grails bubble?

**Reply by gsartori** [reply]
I don’t know, who can answer this question?

**Reply by TheRetroRoot** [reply]
I've used it in a few Micronaut projects (old habits die hard). Outside of Grails and Micronaut, it has never come up.

**Reply by supsup** [reply]
**Just my view, not saying I'm right, I know beggars can't be choosers so I don't expect any of this, these are just ideas since this thread exists.**  


## 1st:

If you really want to hype Grails you should focus on getting Grails to Java 21 at least, as far as I know Grails 7 will support Java 17.  

Why do I think this Grails 7 will only support Java 17: from the docs about Grails 6 :
_The Grails team is actively planning to include Java 17 support in the next major release, ensuring developers can confidently embrace the latest Java advancements when the time comes._

---

## 2nd:
Focus on Java and Kotin first, Groovy last, or make Groovy more about just the Gradle part of Grails. 

## 3rd:
More Guides, especially on cutting edge stuff

https://guides.grails.org/index.html 
 -- last guide was from last year, seen people post example guides in discord but no responses to them. 

Sitemesh (noticed their team is doing updates/upgrades) / Thymeleaf examples of HTMX 

HTMX first class citizen

Spring Security FUD articles -- don't roll your own

FastHTML comparison  -- how to make a DSL to accomplish the same thing in Grails, and be able to make customizations

## 4th:
Micronaut and Grails should be on the same chat server with different channels.  That same server should have Java / Groovy / Kotlin channels etc. OCI Java World or something like that, a great way for OCI to create an ecosystem of developers.    
   ---You submit a guide that is approved you get special flair etc
   ---Could have a bounty board
   ---Security section etc 


----

**Reply by jamesfredley** [reply]
Grails 7 will be based on Spring Framework 6.1 and Spring Boot 3.2 or 3.3, so min Java version will be 17 and Spring supports up to and including Java 22.

**Reply by supsup** [reply]
> Grails 7 will be based on Spring Framework 6.1 and Spring Boot 3.2 or 3.3, so min Java version will be 17 and is most likely compatible up to and including Java 22, since Spring supports those versions.

that will be huge

**Reply by gsartori** [reply]
On Point 2, well… Grails IS Groovy. Groovy has been there for long time and keeps evolving to these days. We can use it dynamically like Python or statically like Java. Groovy is an excellent language. Now the trend is on Kotlin because of Google, but that doesn’t make Groovy less excellent or Kotlin better.

To be honest, Kotlin does the same things that Java does on the JVM, while Groovy does things neither Java nor Kotlin do. 

There is no need to emarginate such a great language in favour of another one that cannot do what Groovy does.

**Reply by supsup** [reply]
Fair enough, was a bit extreme.

**Reply by jamesfredley** [reply]
Gradle is going to be a limiting factor on the maximum supported Java version for Grails 7.  

Gradle 7.6.4 (the latest 7.x version), used on Grails 6.2, supports up to Java 19.  Gradle 8.5 is required for Java 21 support.   The Grails code is not yet Gradle 8.5 compatible and there will be some level of effort to get there.    

https://docs.gradle.org/current/userguide/compatibility.html

**Reply by gsartori** [reply]
@jamesfredley I guess we should target the latest Java and Gradle versions for Grails 7 then. It is a bit unfortunate that Gradle does not have an LTS policy that makes sense, they are forcing everybody to follow their speedy evolution with small to no real reasons to move from one version to the next one. I used to like Gradle for its clean configuration style. Today I just wish that Maven could be configurable with YAML instead of XML and I would switch to it immediately. The build system CANNOT be a budget wasting tool.

**Reply by jamesfredley** [reply]
It would be nice for Gradle to add support for new Java versions without forcing you to address all of the features that were removed or changed with the Gradle latest version.  It is a time suck, especially on a large project with a ton of Gradle config that was created over a long period of time.

There is a difference between the Java version Gradle supports for the build and version Grails will run on, even with 6.2.

**Reply by jamesfredley** [reply]
Several Grails projects have been updated to Gradle 8.10, which supports up to Java 23.  There is still more work remaining, but good progress so far.

**Reply by rlconst** [reply]
We have a multi-module project and grails effectively bottleneck for upgrading gradle which itself is a bottleneck to update another frameworks e.g. micronaut

**Reply by jamesfredley** [reply]
Grails 7 is currently using Gradle 8.10.2.  You can preview the SNAPSHOTs on https://github.com/jamesfredley/grails-website-test

**Reply by snimavat** [reply]
It feels so sad, we had a very thriving community

**Reply by jamesfredley** [reply]
The mojo is rising with Grails 7.  https://grails.org/blog/2024-10-01-grails-community-contributions.html

**Comment by TheRetroRoot** [comment]
Move to another chat system. Since Slack deletes messages after 10k, it's basically deleting the value of the community. I know there are external logging things that can be used. But really, visiting the chat and seeing no history really adds to the "dead community" vibe.

Move to something like Discord which has tools in place to begin rebuilding the community and creates opportunities for engagement. There are some great tools available for Discord which would allow us to begin having virtual conferences, workshops, and more.

**Reply by supsup** [reply]
agreed

**Reply by davydotcom** [reply]
Discord makes a lot of sense to me

**Reply by gsartori** [reply]
I've created a Discord channel if we want to experiment with it, who should be added as administrator?

https://discord.gg/z8uWkqpw

**Reply by TheRetroRoot** [reply]
I manage a few different Discord servers and am happy to pitch in on admin work + setting it up for use.

**Reply by sdelamo** [reply]
I am not sure moving the community to a new chat system is a good idea in this moment of the framework.

**Reply by gsartori** [reply]
@sdelamo it can be energising though, we need energy now, we need “new things”, new people, new energies. We need fresh air to breath again.

**Reply by sbglasius** [reply]
Groovy slack is mirrored here: https://groovy.linen.dev/ and Grails slack is mirrored here: https://grails.linen.dev/ - both with retained history. But I agree that moving away from Slack to any (free) platform would be better.

**Comment by monetschemist** [comment]
I've used Groovy and Grails since 2010 or so.  The only "real" web apps I've built have been with Grails.  Said that, I don't spend my days developing stuff in any language / framework; it's what I do occasionally as the need in my work arises.  Beyond that, I've written about Groovy on https://opensource.com/users/clhermansen and on https://opensource.net/author/chrishermansen/  My comments below about "marketing Grails" come from this perspective.

Things I feel contribute to a sense of "Grails being dying or dead":

1. out-of-date tutorials (sometimes very out of date) e.g. https://guides.grails.org/building-a-react-app/guide/index.html is based on Grails 3.3.5
2. superficial documentation that offers limited explanation, doesn't drill down into more complex examples and doesn't explicitly state best practices e.g. https://gorm.grails.org/latest/hibernate/manual/index.html#gormAssociation which avoids the semantic distinction between many-to-one as a "lookup relationship", using the Nose-Face example (which probably is not going to be a lookup relationship but rather a parent-child relationship), which offers the idea of cascading saves (implicitly for parent-child relationships, but does it also work for "lookup relationships" where there is no "belongs to", etc)
3. lack of up-to-date third party reference material e.g. Burt Beckwith's exceptionally useful _Programming Grails: Best Practices for Experienced Grails Developers_ is based on Grails 2.0, Smith and Ledbrook's _Grails in Action- is based on Grails 2.0, etc.

I get that updating these things requires significant commitment on the part of volunteers (especially writing books), and that in some cases a sense of "dead or dying" is a self-fulfilling prophecy (for example, is O'Reilly or Manning interested in a new book on Grails?).  So this is tough, but I see these three points as a collectively LARGE barrier for new Grails users to overcome.

Besides those three points, I understand the concerns voiced over up-to-date support for IntelliJ and VS Code, as I imagine many (most?) developers either use one of those by choice or are required to by their employers.

The issue of Grails dependencies being behind seems an important one as well.  The whole business of having to stick with Java 8 for so long offered the obvious conclusion that Grails was not keeping up with the times.

What to do about this?  I would be happy to help with updating and enhancing documentation, like tutorials, including preparing a GORM tutorial with more real-world semantics.  I assume such things are addressed with pull requests, but the kinds of changes I propose may run a bit deeper than correcting typos.  Is this a worthwhile thing to undertake?

**Reply by jamesfredley** [reply]
@monetschemist We would love to have your assistance with updating documentation for Grails 7.  Most of the conversations happen on https://slack.grails.org/, if you would like to join.   And I'd be glad to jump on a quick call to get you up to speed and connected.

**Reply by jamesfredley** [reply]
Documentation is in the asciidoc format which lives in https://github.com/grails/grails-doc/tree/7.0.x/src/en and PRs can be submitted for changes.

**Reply by monetschemist** [reply]
Ok, I'm going to pick a tutorial that deals with a topic I feel matches my ability to contribute, then I will review it to see what I don't understand and ask whatever questions I need to clear up on the Slack channel, then make my changes, then submit the PR.  Sound OK?

**Reply by jamesfredley** [reply]
Sounds great.

**Reply by jeffscottbrown** [reply]
> for example, is O'Reilly or Manning interested in a new book on Grails?

I don't know about O'Reilly but both Manning and Apress have expressed interest at times during the years since the most recent Definitive Guide To Grails books were published.

**Reply by monetschemist** [reply]
I  would love to see the Definitive Guide to Grails updated.  I would love to contribute to that.  I feel I am limited by my lack of knowledge of "best practices" in the newer versions of Grails.  For example, all the new (to me at least) annotation stuff in the services layer - I understand the indications given in the official docs but I don't see how that orients me to best practices.  Another example, I feel there must be great utility to command classes, but my current level of knowledge is that they are mostly useful for decoupling validation from domain classes (and I'm not sure why I would want to do that, since I don't see a huge opportunity to share validation between different domain classes).

**Reply by sbglasius** [reply]
There was also a book by @erichelgeson on Grails 3, that you could buy online. Unfortunately that is now taken off-line. Perhaps it could be revitalized? 
The last time it was visible was here: https://web.archive.org/web/20191213200212/https://www.grails3book.com/

**Comment by ewcole** [comment]
I think the real problem is a lack of commitment to backwards compatibility from the developers of Groovy and Grails.  I've written a lot of applications in Grails 2.5, and would be happy to stay with it forever if it weren't a sinking ship.  But subsequent versions of Grails have changed so much that it is like learning a completely new framework, and who has time for that?  This is especially painful if you are trying to do anything beyond the examples on the basic tutorial, which is just when you look for sample code on Stack Overflow or blog posts.  If you didn't break the existing code, then you could keep using the old documentation forever.

In contrast, I could run any PL/SQL code that I've written in the last 20 years without any changes whatsoever.  I never waste time on the squirrel cage when I use that tool.

Forgive me if I sound critical, but I am very frustrated, and I am beginning to think that I made a serious mistake when I bet the farm on Grails.

**Comment by kico-itexto** [comment]
You know something that was a mistake in my opinion? The removal of the
"grails upgrade" command.


Em qua., 20 de nov. de 2024 às 17:53, Ed ***@***.***>
escreveu:

> I think the real problem is a lack of commitment to backwards
> compatibility from the developers of Groovy and Grails. I've written a lot
> of applications in Grails 2.5, and would be happy to stay with it forever
> if it weren't a sinking ship. But subsequent versions of Grails have
> changed so much that it is like learning a completely new framework, and
> who has time for that? This is especially painful if you are trying to do
> anything beyond the examples on the basic tutorial, which is where you look
> for sample code on Stack Overflow or blog posts. If you didn't break the
> existing code, then you could keep using the old documentation forever.
>
> In contrast, I could run any PL/SQL code that I've written in the last 20
> years without any changes whatsoever. I never waste time on the squirrel
> cage when I use that tool.
>
> Forgive me if I sound critical, but I am very frustrated, and I am
> beginning to think that I made a serious mistake when I bet the farm on
> Grails.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/grails/discussions/13559#discussioncomment-11328590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6W7AASVKVGTXUBN5PM4TL2BTZEVAVCNFSM6AAAAABLCQTY2WVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZSHA2TSMA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

Henrique Lobo Weissmann
Diretor Técnico
<https://www.facebook.com/itexto> <https://www.twitter.com/itexto>
<https://www.linkedin.com/company/itexto>
<https://www.instagram.com/itextobr>
*Tel:* +55 (31) 2573-1979 <+55(31)2573-1979>
*Email:* ***@***.***
*Web:* www.itexto.com.br

**Reply by jeffscottbrown** [reply]
FYI... https://github.com/grails/grails-core/issues/13784

**Comment by kico-itexto** [comment]
Wonderfull Jeff!
Good move!

Em qui., 21 de nov. de 2024 às 16:27, Jeff Scott Brown <
***@***.***> escreveu:

> FYI... #13784 <https://github.com/grails/grails-core/issues/13784>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/grails/discussions/13559#discussioncomment-11340499>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6W7AD6JGNID7JUSMFLMPL2BYX2JAVCNFSM6AAAAABLCQTY2WVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZUGA2DSOI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

Henrique Lobo Weissmann
Diretor Técnico
<https://www.facebook.com/itexto> <https://www.twitter.com/itexto>
<https://www.linkedin.com/company/itexto>
<https://www.instagram.com/itextobr>
*Tel:* +55 (31) 2573-1979 <+55(31)2573-1979>
*Email:* ***@***.***
*Web:* www.itexto.com.br
